syntax = "proto3";

package kvi;
option go_package = "github.com/thirawat27/kvi/pkg/grpc;kvi_grpc";

message GetRequest {
    string key = 1;
}

message GetResponse {
    string id = 1;
    string data_json = 2; // JSON representation for dynamic map
}

message PutRequest {
    string key = 1;
    string data_json = 2;
}

message PutResponse {
    bool success = 1;
}

message VectorSearchRequest {
    repeated float vector = 1;
    int32 k = 2;
}

message VectorSearchResponse {
    message Result {
        string id = 1;
        string data_json = 2;
    }
    repeated Result results = 1;
}

service KviService {
    rpc Get(GetRequest) returns (GetResponse);
    rpc Put(PutRequest) returns (PutResponse);
    rpc VectorSearch(VectorSearchRequest) returns (VectorSearchResponse);
}
